src/neurostatslib/glm.py

```diff
@@ -1,431 +1,598 @@
+            Simulate neural activity in response to a feed-forward input
+            using the GLM as a recurrent network.
+        """
+        # check that the model is fitted
+        self._check_is_fit()
+        # extract model params
+        Ws = self.basis_coeff_
+        bs = self.baseline_link_fr_
+
+        (X,) = self._convert_to_jnp_ndarray(X)
+
+        # check input dimensionality
+        self._check_input_dimensionality(X=X)
+        # check consistency between X and params
+        self._check_input_and_params_consistency((Ws, bs), X=X)
+        return self._predict((Ws, bs), X)
 
-                .. math::
-                   \frac{\lambda^k \exp(-\lambda)}{k!}
+    def _score(  # call _negative_log_likelihood
+        self,
+        params: Tuple[jnp.ndarray, jnp.ndarray],
+        X: jnp.ndarray,
+        y: jnp.ndarray,
+    ) -> jnp.ndarray:
+        r"""Score the predicted rates against target neural activity.
 
-                Thus, the negative log of it is:
+        This computes the negative log-likelihood up to a constant term.
 
-                .. math::
-        ¨           -\log{\frac{\lambda^k\exp{-\lambda}}{k!}} &= -[\log(\lambda^k)+\log(\exp{-\lambda})-\log(k!)]
-                   &= -k\log(\lambda)-\lambda+\log(\Gamma(k+1))
+        Note that you can end up with infinities in here if there are zeros in
+        ``predicted_rates``. We raise a warning in that case.
 
-                Because $\Gamma(k+1)=k!$, see
-                https://en.wikipedia.org/wiki/Gamma_function.
+        Parameters
+        ----------
+        params :
+            Values for the spike basis coefficients and bias terms. Shape ((n_neurons, n_features), (n_neurons,)).
+        X :
+            The exogenous variables. Shape (n_time_bins, n_neurons, n_features).
+        y :
+            The target activity to compare against. Shape (n_time_bins, n_neurons).
 
-                And, in our case, ``target_spikes`` is $k$ and
-                ``predicted_firing_rates`` is $\lambda$
+        Returns
+        -------
+        :
+            The model negative log-likehood. Shape (1,).
 
         """
-        x = target_spikes * jnp.log(predicted_firing_rates)
-        # this is a jax jit-friendly version of saying "put a 0 wherever
-        # there's a NaN". we do this because NaNs result from 0*log(0)
-        # (log(0)=-inf and any non-zero multiplied by -inf gives the expected
-        # +/- inf)
-        x = jnp.where(jnp.isnan(x), jnp.zeros_like(x), x)
-        # see above for derivation of this.
-        return jnp.mean(
-            predicted_firing_rates - x + jax.scipy.special.gammaln(target_spikes + 1)
-        )
+        predicted_rate = self._predict(params, X)
+        return self.noise_model.negative_log_likelihood(predicted_rate, y)
+
+    def score(
+        self,
+        X: Union[NDArray, jnp.ndarray],
+        y: Union[NDArray, jnp.ndarray],
+        score_type: Literal["log-likelihood", "pseudo-r2"] = "pseudo-r2",
```

Comment:
honestly, I need to figure out what is wrong with my pseudo-r2, because in the end
$R^2_{\text{pseudo}}= k^2 \cdot  \text{LL} + h$
where k and h are constants and LL is the log-likelihood, so it is basically a re-normalized log-likelihood, so that it is between [0,1] for the training set.

Reply:
there was a line missing in my deviance computation...

Reply:
I did investigate the issue, I was squaring the deviance, which was wrong. What we have implemented now is bug free, the formula is a ratio of deviances (no squaring), and the deviance is compared against statsmodels. If you want to check that the formula matches the implementation I have the references in the docstrings. (The 
McFadden pseudo-r2 matches statsmodels, the Choen doesn't have a corresponding implementation in statsmodels).

My reasoning about the pseudo-r2 being a re-scaled version of the log-likelihood was flawed. the pseudo r^2  depends on the null-model log-likelihood (and that  the saturated model depending on the definition), which are not constant.

(for example, the null model is a Poisson with constant rate equal to the mean firing rate, but if the rate in the test set is for some reason different, then the pseudo r-square will not be linearly related to the log-likelihood).